external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java

```diff
@@ -0,0 +1,566 @@
+        initializeCurator();
+        List<Shard> shards = this.getShards();
+        LOG.info("myTaskIndex is " + myTaskIndex);
+        LOG.info("totalTasks is " + totalTasks);
+        int i = myTaskIndex;
+        while (i < shards.size()) {
+            LOG.info("Shard id " + shards.get(i).getShardId() + " assigned to task " + myTaskIndex);
+            toEmitPerShard.put(shards.get(i).getShardId(), new LinkedList<Record>());
+            i += totalTasks;
+        }
+        initializeFetchedSequenceNumbers();
+        refreshShardIteratorsForNewRecords();
+    }
+
+    void next (SpoutOutputCollector collector) {
+        if (shouldCommit()) {
+            commit();
+        }
+        KinesisMessageId failedMessageId = config.getFailedMessageRetryHandler().getNextFailedMessageToRetry();
+        if (failedMessageId  != null) {
+            // if the retry service returns a message that is not in failed set then ignore it. should never happen
+            BigInteger failedSequenceNumber = new BigInteger(failedMessageId.getSequenceNumber());
+            if (failedPerShard.containsKey(failedMessageId.getShardId()) && failedPerShard.get(failedMessageId.getShardId()).contains(failedSequenceNumber)) {
+                if (!failedandFetchedRecords.containsKey(failedMessageId)) {
+                    fetchFailedRecords(failedMessageId);
+                }
+                if (emitFailedRecord(collector, failedMessageId)) {
+                    failedPerShard.get(failedMessageId.getShardId()).remove(failedSequenceNumber);
+                    config.getFailedMessageRetryHandler().failedMessageEmitted(failedMessageId);
+                    return;
+                } else {
+                    LOG.debug("failedMessageEmitted not called on retrier for " + failedMessageId + ". This can happen a few times but should not happen " +
+                            "infinitely");
+                }
+            } else {
+                LOG.debug("failedPerShard does not contain " + failedMessageId + ". This should never happen.");
+            }
+        }
+        LOG.debug("No failed record to emit for now. Hence will try to emit new records");
+        // if maximum uncommitted records count has reached, so dont emit any new records and return
+        if (!(getUncommittedRecordsCount() < config.getMaxUncommittedRecords())) {
+            LOG.debug("maximum uncommitted records count has reached. so not emitting any new records and returning");
+            return;
+        }
+        // early return as no shard is assigned - probably because number of executors > number of shards
+        if (toEmitPerShard.isEmpty()) {
+            LOG.debug("No shard is assigned to this task. Hence not emitting any tuple.");
+            return;
+        }
+
+        if (shouldFetchNewRecords()) {
+            fetchNewRecords();
+        }
+        emitNewRecord(collector);
+    }
+
+    void ack (KinesisMessageId kinesisMessageId) {
```

Comment:
I documented the logic in the commit method since thats where it writes to zookeeper. Let me know if that is not enough or if i should move it to somewhere else.